symengine/lambda_double.h

```diff
@@ -0,0 +1,309 @@
+#ifndef SYMENGINE_LAMBDA_DOUBLE_H
+#define SYMENGINE_LAMBDA_DOUBLE_H
+
+#include <cmath>
+#include <functional>
+#include <complex>
+
+#include <symengine/basic.h>
+#include <symengine/symbol.h>
+#include <symengine/add.h>
+#include <symengine/integer.h>
+#include <symengine/rational.h>
+#include <symengine/complex.h>
+#include <symengine/mul.h>
+#include <symengine/pow.h>
+#include <symengine/functions.h>
+#include <symengine/constants.h>
+#include <symengine/visitor.h>
+#include <symengine/eval_double.h>
+
+namespace SymEngine {
+
+template<typename T, typename U>
+class LambdaDoubleVisitor : public BaseVisitor<U> {
+protected:
+/*
+   The 'result_' variable is assigned into at the very end of each visit()
+   methods below. The only place where these methods are called from is the
+   line 'b.accept(*this)' in apply() and the 'result_' is immediately
+   returned. Thus no corruption can happen and apply() can be safely called
+   recursively.
+*/
+
+    typedef std::function<T(const std::vector<T> &x)> fn;
+    fn result_;
+    vec_basic symbols;
+public:
+    LambdaDoubleVisitor(U *p) : BaseVisitor<U>(p) {
+    }
+
+    void init(const vec_basic &x, const Basic &b) {
+        symbols = x;
+        apply(b);
+    }
+
+    fn apply(const Basic &b) {
+        b.accept(*this);
+        return result_;
+    }
+
+    T call(const std::vector<T> &vec) {
+        return result_(vec);
+    }
+
+    void bvisit(const Integer &x) {
+        T tmp = x.i.get_d();
+        result_ = [=](const std::vector<T> &x_){ return tmp; };
+    }
+
+    void bvisit(const Rational &x) {
+        T tmp = x.i.get_d();
+        result_ = [=](const std::vector<T> &x){ return tmp; };
+    }
+
+    void bvisit(const RealDouble &x) {
+        T tmp = x.i;
+        result_ = [=](const std::vector<T> &x){ return tmp; };
+    }
+
+#ifdef HAVE_SYMENGINE_MPFR
+    void bvisit(const RealMPFR &x) {
+        T tmp = mpfr_get_d(x.i.get_mpfr_t(), MPFR_RNDN);
+        result_ = [=](const std::vector<T> &x){ return tmp; };
+    }
+#endif
+
+    void bvisit(const Add &x) {
+        fn tmp = apply(*x.coef_);
+        fn tmp1, tmp2;
+        for (auto &p: x.dict_) {
+            tmp1 = apply(*(p.first));
+            tmp2 = apply(*(p.second));
+            tmp = [=](const std::vector<T> &x){ return tmp(x) + tmp1(x) * tmp2(x); };
```

Comment:
Here's an article about how to capture by move a large vector. https://marcoarena.wordpress.com/2012/11/01/learn-how-to-capture-by-move/